### PR TITLE
applications: serial_lte_modem: improve HTTPC service

### PR DESCRIPTION
**Pull request by lats1980**
This PR improves HTTPC service with below 2 commits:

1. applications: serial_lte_modem: integrate HTTPC to SLM data mode
2. applications: serial_lte_modem: fix infinite loop with no exit

### CODE REVIEW

File: applications/serial_lte_modem/doc/HTTPC_AT_commands.rst [hunk 1]
```diff
@@ -128,9 +128,10 @@ Syntax
   Each header field should end with ``<CR><LF>``.
 * The ``<payload_length>`` is an integer.
   It represents the length of the payload.
-  If ``payload_length`` is greater than ``0``, the SLM will enter the pass-through mode and expect the upcoming UART input data as payload.
+  If ``payload_length`` is greater than ``0``, the SLM will enter the data mode and expect the upcoming UART input data as payload.
```
Comment (ru-fu) [inline applications/serial_lte_modem/doc/HTTPC_AT_commands.rst:131]:
```suggestion
  If ``payload_length`` is greater than ``0``, the SLM will enter data mode and expect the upcoming UART input data as payload.
```

File: applications/serial_lte_modem/doc/HTTPC_AT_commands.rst [hunk 2]
```diff
@@ -128,9 +128,10 @@ Syntax
...
   The SLM will then send the payload to the HTTP server until the ``payload_length`` bytes are sent.
-  To abort sending the payload, use ``AT#XHTTPCCON=0`` to disconnect from the server.
+  To abort sending the payload, use ``CONFIG_SLM_DATAMODE_TERMINATOR`` to terminate data mode. The default pattern string is "+++".
+  UART silence ``CONFIG_SLM_DATAMODE_SILENCE`` before and after the pattern string is used to exit data mode.
```
Comment (ru-fu) [inline applications/serial_lte_modem/doc/HTTPC_AT_commands.rst:134]:
```suggestion
  To abort sending the payload, terminate data mode by sending the terminator string defined in :option:`CONFIG_SLM_DATAMODE_TERMINATOR`.
  The default pattern string is "+++".
  Keep in mind that UART silence as configured in :option:`CONFIG_SLM_DATAMODE_SILENCE` is required before and after the pattern string.
```

File: applications/serial_lte_modem/doc/HTTPC_AT_commands.rst [hunk 3]
```diff
@@ -207,48 +208,53 @@ Syntax
...
+  * ``0`` - There is more HTTP response data to come.
+  * ``1`` - The entire HTTP response has been received.
 
 * The ``<response>`` is the raw data of the HTTP response, including headers and body.
 
 Example
 ~~~~~~~
 
-The following example sends a POST request to send data to the server with an optional header.
+The following example sends a PUT request to send JSON format data to the server with an optional header.
```
Comment (ru-fu) [inline applications/serial_lte_modem/doc/HTTPC_AT_commands.rst:219]:
```suggestion
The following example sends a PUT request to send data in JSON format to the server, with an optional header.
```

File: applications/serial_lte_modem/doc/HTTPC_AT_commands.rst [hunk 4]
```diff
@@ -207,48 +208,53 @@ Syntax
...
 
 ::
 
-   AT#XHTTPCREQ="POST","/post","User-Agent: SLM/1.2.0
-   Accept: */*
-   Content-Type: application/x-www-form-urlencoded
-   Content-Length: 20
-   ",20
+   at#xhttpccon=1,"iot.cht.com.tw",80
```
Comment (ru-fu) [inline applications/serial_lte_modem/doc/HTTPC_AT_commands.rst:223]:
```suggestion
   AT#XHTTPCCON=1,"iot.cht.com.tw",80
```

Comment (ru-fu) [reply applications/serial_lte_modem/doc/HTTPC_AT_commands.rst:223]: Can we use example.com as URL?

**Review by ru-fu — CHANGES_REQUESTED** [review]

File: applications/serial_lte_modem/src/http_c/slm_at_httpc.c [hunk 5]
```diff
@@ -337,57 +336,92 @@ static int headers_cb(int sock, struct http_request *req, void *user_data)
...
+		pl_to_send = len;
+	}
+
+	/* Start to send payload */
+	while (pl_sent < pl_to_send) {
+		ssize_t ret;
+
+		ret = send(httpc.fd, data + pl_sent,
+				MIN(pl_to_send - pl_sent,
+				HTTPC_FRAG_SIZE), 0);
```
Comment (rlubos) [inline applications/serial_lte_modem/src/http_c/slm_at_httpc.c:364]: nit: alignment
```suggestion
		ret = send(httpc.fd, data + pl_sent,
			   MIN(pl_to_send - pl_sent, HTTPC_FRAG_SIZE), 0);
```

File: applications/serial_lte_modem/src/http_c/slm_at_httpc.c [hunk 6]
```diff
@@ -618,48 +660,46 @@ int handle_at_httpc_request(enum at_cmd_type cmd_type)
...
 				return err;
 			}
 		}
-		/* start sending request */
-		k_sem_give(&http_req_sem);
+		/* start http request thread */
+		k_thread_create(&httpc_thread, httpc_thread_stack,
+			K_THREAD_STACK_SIZEOF(httpc_thread_stack),
+			httpc_thread_fn, NULL, NULL, NULL,
+			HTTPC_THREAD_PRIORITY, K_USER, K_NO_WAIT);
```
Comment (rlubos) [inline applications/serial_lte_modem/src/http_c/slm_at_httpc.c:702]: nit: alignment
```suggestion
		k_thread_create(&httpc_thread, httpc_thread_stack,
				K_THREAD_STACK_SIZEOF(httpc_thread_stack),
				httpc_thread_fn, NULL, NULL, NULL,
				HTTPC_THREAD_PRIORITY, K_USER, K_NO_WAIT);
```

**Review by rlubos — COMMENTED** [review]
Just some style nits

File: applications/serial_lte_modem/doc/HTTPC_AT_commands.rst (hunk 2 above)
Comment (lats1980) [reply applications/serial_lte_modem/doc/HTTPC_AT_commands.rst]: The description was changed per suggestion.

File: applications/serial_lte_modem/doc/HTTPC_AT_commands.rst (hunk 3 above)
Comment (lats1980) [reply applications/serial_lte_modem/doc/HTTPC_AT_commands.rst]: The description was changed per suggestion.

File: applications/serial_lte_modem/doc/HTTPC_AT_commands.rst (hunk 4 above)
Comment (lats1980) [reply applications/serial_lte_modem/doc/HTTPC_AT_commands.rst]: Yes, example.com  would be more suitable. I've modified it, thanks.

File: applications/serial_lte_modem/src/http_c/slm_at_httpc.c (hunk 5 above)
Comment (lats1980) [reply applications/serial_lte_modem/src/http_c/slm_at_httpc.c]: The code was changed per suggestion, thank you.

File: applications/serial_lte_modem/src/http_c/slm_at_httpc.c (hunk 6 above)
Comment (lats1980) [reply applications/serial_lte_modem/src/http_c/slm_at_httpc.c]: The code was changed per suggestion, thank you.

**Comment by junqingzou** [comment]
One general comment, when HTTPC client enters data mode, you cannot send any URC

File: applications/serial_lte_modem/src/http_c/slm_at_httpc.c [hunk 7]
```diff
@@ -337,57 +336,91 @@ static int headers_cb(int sock, struct http_request *req, void *user_data)
...
+			return -errno;
+		}
+		LOG_DBG("send %d bytes payload", ret);
+		pl_sent += ret;
+		httpc.total_sent += ret;
+	}
+
+	if (httpc.total_sent == httpc.pl_len) {
+		LOG_DBG("Successfully send %d bytes payload", httpc.total_sent);
+		k_sem_give(&http_req_sem);
```
Comment (junqingzou) [inline applications/serial_lte_modem/src/http_c/slm_at_httpc.c:377]: Is it correct to complete with one datamode send only? I'd prefer to see that normal ending by "+++" and explicitly end in data mode callback.

File: applications/serial_lte_modem/src/http_c/slm_at_httpc.c [hunk 8]
```diff
@@ -337,57 +336,91 @@ static int headers_cb(int sock, struct http_request *req, void *user_data)
...
+	while (pl_sent < pl_to_send) {
+		ssize_t ret;
+
+		ret = send(httpc.fd, data + pl_sent,
+			   MIN(pl_to_send - pl_sent, HTTPC_FRAG_SIZE), 0);
+		if (ret < 0) {
+			LOG_ERR("Fail to send payload: %d", ret);
+			httpc.total_sent = -errno;
+			k_sem_give(&http_req_sem);
+			return -errno;
```
Comment (junqingzou) [inline applications/serial_lte_modem/src/http_c/slm_at_httpc.c:368]: Normally error quit of data mode would need a URC to notify. For example:
		if (exit_datamode()) {
			sprintf(rsp_buf, "\r\n#XTCPCLI: 0,\"datamode\"\r\n");
			rsp_send(rsp_buf, strlen(rsp_buf));
		}

**Review by rlubos — APPROVED** [review]

**Review by ru-fu — APPROVED** [review]

File: applications/serial_lte_modem/src/http_c/slm_at_httpc.c (hunk 8 above)
Comment (lats1980) [reply applications/serial_lte_modem/src/http_c/slm_at_httpc.c:368]: When "+++" terminates data mode, URC "#XHTTPCREQ: 0" is notified. If send/recv error "#XHTTPCREQ: <errno>" is notified.

File: applications/serial_lte_modem/src/http_c/slm_at_httpc.c (hunk 7 above)
Comment (lats1980) [reply applications/serial_lte_modem/src/http_c/slm_at_httpc.c:377]: If data length is reached, there is no need to send more data. Normally length is specified in the HTTP request header "Content-Length"

**Comment by lats1980** [comment]
> One general comment, when HTTPC client enters data mode, you cannot send any URC. Also could you rename "do_httpc_send_datamode" to be "httpc_datamode_callback" so as to align with other modules?

Now "do_httpc_send_datamode" is renamed as "httpc_datamode_callback".

As for the URC, HTTP client works differently than data mode in other service. HTTP works in REQUEST-RESPONSE manner. In SLM, datamode only applies to REQUEST (uplink). Therefore sending URC won't be mixed with RESPONSE URC.

**Review by junqingzou — APPROVED** [review]
Thanks for explanation of the HTTP view of data mode.
Understood well now.

**Pull request merged**